### PR TITLE
feat(protocol-designer): highlight tips per substep

### LINKED ISSUE #2537: On deck display tip user per row of a step
As a user I'd like to be clear on how tips are used during my step

## Acceptance criteria
- When a user hovers on a row of a step only the tip used in the step is highlighted

### PR DESCRIPTION
**Pull request by IanLondon**
## overview

Closes #2537

### Testing

There's a large surface to test across:

* Should work with transfer, distribute, and consolidate step types
* Should work with both single and multi-channel pipettes
* Should work with all change tip options

Also, step-level tip highlighting should work exactly as before (eg hover mouse over title of step in step list, not on a substep, should show all tips used in step)

Code is getting ugly because everything in substeps has to happen almost the same but slightly differently across single vs multi-channel pipettes, I'm open to suggestions about how to make these functions cleaner

### CODE REVIEW

File: protocol-designer/src/top-selectors/substeps.js [hunk 1]
```diff
@@ -30,8 +30,9 @@ export const allSubsteps: Selector<AllSubsteps> = createSelector(
     robotStateTimeline,
     _initialRobotState,
   ) => {
+    const timeline = [{robotState: _initialRobotState}, ...robotStateTimeline.timeline]
```
Comment (IanLondon) [inline protocol-designer/src/top-selectors/substeps.js]: this timeline was being created new inside each loop of the `reduce`, I think it was just a mistake?

File: protocol-designer/src/steplist/substepTimeline.js [hunk 2]
```diff
@@ -1,47 +1,83 @@
...
+        return {...acc, errors: nextFrame.errors}
       }
+      prevRobotState = nextFrame.robotState
 
       // NOTE: only aspirate and dispense commands will appear alone in atomic commands
       // from compound command creators (e.g. transfer, distribute, etc.)
-      const isAtomic = nextFrame.commands.length === 1
-      const commandGroup = nextFrame.commands[0]
-      if (isAtomic && (commandGroup.command === 'aspirate' || commandGroup.command === 'dispense')) {
+      const firstCommand = nextFrame.commands[0]
```
Comment (IanLondon) [inline protocol-designer/src/steplist/substepTimeline.js]: `commandGroup` is really a single `Command` object not a group, so I renamed this for multi and single

File: protocol-designer/src/top-selectors/substeps.js (hunk 1 above)
Comment (b-cooper) [reply protocol-designer/src/top-selectors/substeps.js]: Good catch

File: protocol-designer/src/top-selectors/tip-contents/index.js [hunk 3]
```diff
@@ -32,16 +33,16 @@ function getTipHighlighted (
     if (c.command === 'pick-up-tip' && c.params.labware === labwareId) {
       const commandWellName = c.params.well
       const pipetteId = c.params.pipette
-      const labwareName = StepGeneration.getLabwareType(labwareId, robotState)
+      const labwareType = StepGeneration.getLabwareType(labwareId, robotState)
       const channels = StepGeneration.getPipetteChannels(pipetteId, robotState)
 
-      if (!labwareName) {
+      if (!labwareType) {
         console.error(`Labware ${labwareId} missing labwareName. Could not get tip highlight state`)
```
Comment (b-cooper) [inline protocol-designer/src/top-selectors/tip-contents/index.js]: This error is a little confusing now, could use some rewording.

**Pull request merged**